crates/in_stream/src/tcp.rs

```diff
@@ -0,0 +1,323 @@
+fn tcp_url_to_socket_addr(url: &Url2) -> Result<std::net::SocketAddr> {
+    if url.scheme() != SCHEME || url.host_str().is_none() || url.port().is_none() {
+        return Err(Error::new(
+            ErrorKind::InvalidInput,
+            format!("got: '{}', expected: '{}://host:port'", SCHEME, url),
+        ));
+    }
+    let rendered = format!("{}:{}", url.host_str().unwrap(), url.port().unwrap());
+    match rendered.parse() {
+        Ok(addr) => Ok(addr),
+        Err(_) => Err(Error::new(
+            ErrorKind::InvalidInput,
+            format!("could not parse '{}', as 'host:port'", rendered),
+        )),
+    }
+}
+
+#[derive(Debug)]
+/// configuration options for the listener bind call
+pub struct TcpBindConfig {
+    pub backlog: i32,
+}
+
+impl Default for TcpBindConfig {
+    fn default() -> Self {
+        Self { backlog: 1024 }
+    }
+}
+
+impl InStreamConfig for TcpBindConfig {}
+
+/// basic tcp socket server/listener
+#[derive(Debug)]
+pub struct InStreamListenerTcp(pub std::net::TcpListener);
+
+impl InStreamListenerTcp {
+    pub fn bind(url: &Url2, config: TcpBindConfig) -> Result<Self> {
+        InStreamListenerTcp::raw_bind(url, config)
+    }
+}
+
+impl InStreamListener<&mut [u8], &[u8]> for InStreamListenerTcp {
+    type Stream = InStreamTcp;
+
+    fn raw_bind<C: InStreamConfig>(url: &Url2, config: C) -> Result<Self> {
+        let config = TcpBindConfig::from_gen(config)?;
+        let addr = tcp_url_to_socket_addr(url)?;
+        let listener = net2::TcpBuilder::new_v4()?
+            .bind(addr)?
+            .listen(config.backlog)?;
+        listener.set_nonblocking(true)?;
+        Ok(Self(listener))
+    }
+
+    fn binding(&self) -> Url2 {
+        let local = self.0.local_addr().unwrap();
+        Url2::parse(&format!("{}://{}:{}", SCHEME, local.ip(), local.port()))
+    }
+
+    fn accept(&mut self) -> Result<<Self as InStreamListener<&mut [u8], &[u8]>>::Stream> {
+        let (stream, _addr) = self.0.accept()?;
+        stream.set_nonblocking(true)?;
+        InStreamTcp::priv_new(stream, None)
+    }
+}
+
+impl InStreamListenerStd for InStreamListenerTcp {
+    type StreamStd = InStreamTcp;
+
+    fn accept_std(&mut self) -> Result<<Self as InStreamListenerStd>::StreamStd> {
+        self.accept()
+    }
+}
+
+#[derive(Debug)]
+/// configuration options for tcp connect
+pub struct TcpConnectConfig {
+    pub connect_timeout_ms: Option<u64>,
+}
+
+impl Default for TcpConnectConfig {
+    fn default() -> Self {
+        Self {
+            connect_timeout_ms: Some(5000),
+        }
+    }
+}
+
+impl InStreamConfig for TcpConnectConfig {}
+
+#[derive(Debug)]
+struct TcpConnectingData {
+    addr: std::net::SocketAddr,
+    connect_timeout: Option<std::time::Instant>,
+}
+
+/// basic tcp socket stream
+#[derive(Shrinkwrap, Debug)]
+#[shrinkwrap(mutable)]
+pub struct InStreamTcp {
+    #[shrinkwrap(main_field)]
+    pub stream: std::net::TcpStream,
+    connecting: Option<TcpConnectingData>,
```

Comment:
I'm surprised an optional enum is enough here, opposed to a custom enum with lots of various state transitions.

Reply:
yeah, the [net2](https://crates.io/crates/net2) non-blocking connect api is a little odd. It specifies just calling the same connect call repeatedly on the same stream until it gives something other than a WouldBlock response.